### PR TITLE
ENH: make sparse `sum` cast dtypes like NumPy `sum` for 32-bit Python.

### LINKED ISSUE #5972: Test failures for sparse sum tests on 32-bit Python
Getting a large amount of test failures like the one below, I assume that this is due to gh-5907:

    ======================================================================
    FAIL: test_base.TestCSC.test_sum(<type 'numpy.uint64'>, 1)
    ----------------------------------------------------------------------
    Traceback (most recent call last):
      File "/usr/lib/python2.7/dist-packages/nose/case.py", line 197, in runTest
        self.test(*self.arg)
      File "/home/rgommers/Code/scipy/scipy/sparse/tests/test_base.py", line 767, in check
        assert_equal(dat.sum(axis=0).dtype, datsp.sum(axis=0).dtype)
      File "/home/rgommers/.local/lib/python2.7/site-packages/numpy/testing/utils.py", line 375, in assert_equal
        raise AssertionError(msg)
    AssertionError: 
    Items are not equal:
     ACTUAL: dtype('uint64')
     DESIRED: dtype('uint32')

And also a large set of overflow warnings, all on the same line:

    /home/rgommers/Code/scipy/scipy/sparse/dok.py:370: RuntimeWarning: overflow encountered in ulonglong_scalars
      result[i] += v * other[j]

### PR DESCRIPTION
**Pull request by Tiepies**
This fix should ensure that sparse matrix sum behaves like NumPy dense
matrix sum in terms of dtype casting for 32-bit Python as well as
64-bit Python. Should fix #5972.

### CODE REVIEW

File: scipy/sparse/tests/test_base.py [hunk 1]
```diff
@@ -758,19 +758,22 @@ def test_sum(self):
...
-            assert_equal(dat.sum(axis=None).dtype, datsp.sum(axis=None).dtype)
-            assert_array_almost_equal(dat.sum(axis=0), datsp.sum(axis=0))
-            assert_equal(dat.sum(axis=0).dtype, datsp.sum(axis=0).dtype)
-            assert_array_almost_equal(dat.sum(axis=1), datsp.sum(axis=1))
-            assert_equal(dat.sum(axis=1).dtype, datsp.sum(axis=1).dtype)
-            assert_array_almost_equal(dat.sum(axis=-2), datsp.sum(axis=-2))
-            assert_equal(dat.sum(axis=-2).dtype, datsp.sum(axis=-2).dtype)
-            assert_array_almost_equal(dat.sum(axis=-1), datsp.sum(axis=-1))
-            assert_equal(dat.sum(axis=-1).dtype, datsp.sum(axis=-1).dtype)
+            with np.errstate(over='ignore'):
```
Comment (Tiepies) [inline scipy/sparse/tests/test_base.py]: The context manager is to avoid the integer overflow warnings that are thrown when summing over a downcasted matrix. I don't know if squelching the warnings like this is appropriate, though.

**Comment by perimosocordiae** [comment]
LGTM. I think the overflow warnings are useful to end users, so I'm fine with just ignoring them in the tests.

File: scipy/sparse/tests/test_base.py (hunk 1 above)
Comment (Eric89GXL) [inline scipy/sparse/tests/test_base.py]: Do all of these lines emit warnings, or only some of them? It's best only to catch the ones that do, and add a comment saying what's being ignored (in this case, expected overflows)

Comment (perimosocordiae) [inline scipy/sparse/tests/test_base.py]: All of them emit warnings, but only for particular `dtype` arguments. I suppose we could make the suppression conditional on `dtype`:

    with np.errstate(over=('ignore' if np.dtype(dtype).itemsize < 4 else 'warn')):
        ...

But I'm not convinced that it's worth the extra complexity.

Comment (Tiepies) [inline scipy/sparse/tests/test_base.py]: If you only want to catch the tests that throw overflow warnings in the context manager, you need something like this:
````
# Ignore the overflow warnings that result from casting negative
# numbers to unsigned integers and then using the dok_matrix
# implementation of matrix * vector multiplication:
with np.errstate(over='ignore' if (
        np.any(matrices[j] < 0) and
        np.dtype(dtype).kind == 'u' and
        datsp.getformat() == 'dok' )
                 else 'warn'):
    ...
````
This might be overkill, though. Maybe I could leave it as it is, and just add the comment? An alternative is to remove the test cases in which negative numbers are cast to unsigned integers. That would assume that those cases do not need to behave consistently with NumPy. If the casting behavior from int to uint is assumed to be defined, however, then removal of those cases is also not desirable, I think.

**Comment by pv** [comment]
thanks, merged
